# 28일차 (2024-06-14)

## 에러
- JVM에서 발생
- 프로그램의 비정상적인 종료
- 프로그래머가 직접 처리할 수 없다.
- ex) `OutOfMemoryError`, `StackOverflowError`

## 예외
- 프로그램의 비정상적인 상황
- 프로그래머가 적절히 처리할 수 있다.
- ex) `NullPointerException`, `ArrayIndexOutOfBoundsException`

```java
       Throwable // 모든 에러와 예외의 최상위 클래스
         /   \
        /     \
     Error  Exception // 개발자가 처리할 수 있는 예외 상황
              /   \
             /     \
  RuntimeException  IOException
        /  \           /  \
       /    \         /    \
 NullPointer  ...  FileNotFound  ...
 Exception         Exception
```
### 언체크드 예외 (RuntimeException)
- 컴파일 시점에서 예외 처리를 강제하지 않음.
- `RuntimeException` 하위 예외

### 체크드 예외
- 컴파일 시점에서 예외 처리를 강제함.
- `RuntimeException` 예외를 제외한 `Exception` 하위 예외

UncheckedException.java
```java
public class UncheckedException extends RuntimeException {
    public UncheckedException(String message) {
        super(message);
    }
}
```
CheckedException.java
```java
public class CheckedException extends Exception {
    public CheckedException(String message) {
        super(message);
    }
}
```

ExceptionApplication.java
```java
public class ExceptionApplication {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int a = sc.nextInt();
        int b = sc.nextInt();
        try {
            System.out.println(divide1(a, b)); // 언체크드 예외
        } catch (UncheckedException e) {
            System.out.println(e.getMessage());
        }

        try {
            System.out.println(divide2(a, b)); // 체크드 예외
        } catch (CheckedException e) {
            System.out.println(e.getMessage());
        } finally {
            sc.close(); // 예외 발생 여부와 관계없이 실행
        }
    }

    public static double divide1(int a, int b) {
        if (b == 0) throw new UncheckedException("0으로 나눌 수 없습니다.");
        return (double) a / b;
    }

    public static double divide2(int a, int b) throws CheckedException {
        if (b == 0) throw new CheckedException("0으로 나눌 수 없습니다.");
        return (double) a / b;
    }
}
```
출력 결과
```
0으로 나눌 수 없습니다.
0으로 나눌 수 없습니다.
```

---

### 요약
- 에러는 프로그래머가 처리 불가
- 예외는 프로그래머가 처리 가능
- 에러와 예외의 최상위 클래스는 Throwable
- 언체크드 예외는 컴파일 시점에서 예외 처리 필수가 아니다.
  - 예외 처리를 해야할 경우 `try-catch` 블록 사용
- 체크드 예외는 컴파일 시점에서 예외 처리가 필수이다.
  - 메소드에서 `throws`로 명시 필수
  - `catch` 블록에서 처리 필수
- `finally` 블록은 예외 발생 여부와 상관없이 실행


